samza-azure/src/main/java/org/apache/samza/system/eventhub/consumer/EventHubEntityConnection.java

```diff
@@ -0,0 +1,132 @@
+/*
+* Licensed to the Apache Software Foundation (ASF) under one
+* or more contributor license agreements.  See the NOTICE file
+* distributed with this work for additional information
+* regarding copyright ownership.  The ASF licenses this file
+* to you under the Apache License, Version 2.0 (the
+* "License"); you may not use this file except in compliance
+* with the License.  You may obtain a copy of the License at
+*
+*   http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing,
+* software distributed under the License is distributed on an
+* "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+* KIND, either express or implied.  See the License for the
+* specific language governing permissions and limitations
+* under the License.
+*/
+
+package org.apache.samza.system.eventhub.consumer;
+
+import com.microsoft.azure.eventhubs.PartitionReceiveHandler;
+import com.microsoft.azure.eventhubs.PartitionReceiver;
+import com.microsoft.azure.servicebus.ServiceBusException;
+import com.microsoft.azure.servicebus.StringUtil;
+import org.apache.samza.SamzaException;
+import org.apache.samza.system.eventhub.EventHubClientFactory;
+import org.apache.samza.system.eventhub.EventHubClientWrapper;
+import org.apache.samza.system.eventhub.EventHubConfig;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.time.Instant;
+import java.util.Map;
+import java.util.TreeMap;
+
+public class EventHubEntityConnection {
+  private static final Logger LOG = LoggerFactory.getLogger(EventHubEntityConnection.class);
+  final Map<Integer, String> offsets = new TreeMap<>();
+  final Map<Integer, PartitionReceiveHandler> handlers = new TreeMap<>();
+  private final String namespace;
+  private final String entityPath;
+  private final String sasKeyName;
+  private final String sasKey;
+  private final String consumerName;
+  private final EventHubConfig config;
+  private final Map<Integer, PartitionReceiver> receivers = new TreeMap<>();
```

Comment:
Currently, the `EventHubEntityConnection` class has some stateful/mutating operations -
including, some synchronized methods:
```
+  synchronized void addPartition(int partitionId, String offset, PartitionReceiveHandler handler) {}
+  synchronized void connectAndStart() {}
+  synchronized void stop() {}
```
Can we consider separating these and making the class immutable:
```
 +  private final String namespace;
 +  private final String entityPath;
 +  private final String sasKeyName;
 +  private final String sasKey;
 +  private final String consumerName;
 +  private final EventHubConfig config;
```
Also, I'd remove these state-ful fields from this class:
```
 +  private final Map<Integer, PartitionReceiver> receivers = new TreeMap<>();
 +  private EventHubClientWrapper ehClientWrapper;
 +  private final EventHubClientFactory eventHubClientFactory = new EventHubClientFactory();
```

Do you think we can move this map of partition -> PartitionReceiver instances into the Consumer itself since all of these are related to receiving events? Another nice thing is that you can get rid off the `EventHubEntitityConnectionFactory` completely for wiring these stateful components

Reply:
Refactored in order to remove `EventHubConnectionEnitity` https://github.com/apache/samza/pull/308/commits/1c1677faf360d1287717abd4c3ceba8776d365eb